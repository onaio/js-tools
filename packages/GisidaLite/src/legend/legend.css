.legend {
  position: absolute;
  left: 20px;
  color: #555;
  font-size: 11px;
  font-family: 'Montserrat', sans-serif;
  line-height: 1.2;
  z-index: 1;
  margin-right: 75%;
  width: 22%;
  bottom: 0px;
  background: transparent;
  margin-right: 75%;
  bottom: 0;
}
.legend-fill {
  padding: 15px;
  position: relative;
  height: 17px;
  margin: 3px 0;
}
.legend-fill ul {
  padding: 0;
  list-style-type: none;
}
.legend-fill ul li {
  width: 33px;
  float: left;
  display: block;
  color: #fff;
  opacity: 0.8;
}
.legend-label {
  height: 17px;
}
.legend-label li {
  text-align: center;
  height: 17px;
  padding-top: 2px;
}
.legend > span {
  display: inline-block;
  padding-left: 15px;
  padding-top: 5px;
}
