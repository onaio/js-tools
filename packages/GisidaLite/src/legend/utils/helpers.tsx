import React from 'react';

export const legendBlockBuilder = (color: string, width: string, label: string) => {
  return (
    <li key={color} style={{ background: color, width: `${width}%` }}>
      {label}
    </li>
  );
};
