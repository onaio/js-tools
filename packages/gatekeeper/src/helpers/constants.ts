export const OAUTH2_CALLBACK_ERROR = 'oAuth service oauth2Callback failed, data not returned';
export type OAUTH2_CALLBACK_ERROR = typeof OAUTH2_CALLBACK_ERROR;

export const OAUTH2_HTTP_ERROR = 'oAuth service oauth2Callback failed, HTTP status';
export type OAUTH2_HTTP_ERROR = typeof OAUTH2_HTTP_ERROR;

export const GENERIC_ERROR = 'Something went wrong';
export type GENERIC_ERROR = typeof GENERIC_ERROR;

export const LOGIN_URL = '/login';
export type LOGIN_URL = typeof LOGIN_URL;

export const OAUTH_LOGIN_PROMPT = 'Please log in with one of the following providers';
export type OAUTH_LOGIN_PROMPT = typeof OAUTH_LOGIN_PROMPT;

export const NO_PROVIDERS = 'No Providers';
export type NO_PROVIDERS = typeof NO_PROVIDERS;

export const NOTHING_HERE = 'Nothing here!';
export type NOTHING_HERE = typeof NOTHING_HERE;

export const AN_ERROR_OCCURRED = 'An error occurred!';
export type AN_ERROR_OCCURRED = typeof AN_ERROR_OCCURRED;

export const PLEASE_WAIT = 'Please wait...';
export type PLEASE_WAIT = typeof PLEASE_WAIT;

export const WELCOME_BACK = 'Welcome back, ';
export type WELCOME_BACK = typeof WELCOME_BACK;

export const TOKEN_REFRESH_FAILED = 'Failed to refresh token';
export type TOKEN_REFRESH_FAILED = typeof TOKEN_REFRESH_FAILED;
