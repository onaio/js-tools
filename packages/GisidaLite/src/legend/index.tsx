import React from 'react';
import './legend.css';
import { legendBlockBuilder } from './utils/helpers';

/** legend data type */
export interface LegendBlock {
  color: string;
  fillWidth: string;
  label: string;
}
/** Legend props */
export interface LegendProps {
  legendData: LegendBlock[];
  legendCreditText: string;
}

/** Default props for GisidaLite */
export const defaultLegendProps: LegendProps = {
  legendCreditText: 'Age break down',
  legendData: [
    { color: '#bdd7e7', fillWidth: '33px', label: '0' },
    { color: '#4292c6', fillWidth: '33px', label: '1-2' },
    { color: '#2171b5', fillWidth: '33px', label: '3-6' },
    { color: '#08519c', fillWidth: '33px', label: '6-11' },
    { color: '#08306b', fillWidth: '33px', label: '12-16' }
  ]
};
const Legend = (props: LegendProps) => {
  const { legendData, legendCreditText } = props;
  const background: JSX.Element[] = [];
  legendData.forEach((legendBlock: LegendBlock, index: number) => {
    background.push(
      legendBlockBuilder(legendBlock.color, legendBlock.fillWidth, legendBlock.label)
    );
  });
  return (
    <div className="legend">
      <div className="legend-fill legend-label">
        <ul id="legend-background">{background}</ul>
      </div>
      <span>{legendCreditText}</span>
    </div>
  );
};
Legend.defaultProps = defaultLegendProps;

export { Legend };
