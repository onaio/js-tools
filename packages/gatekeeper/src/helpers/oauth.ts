/* eslint-disable camelcase */
import { SessionState } from '@onaio/session-reducer';
import ClientOAuth2, { Options } from 'client-oauth2';
import { OAUTH2_CALLBACK_ERROR } from './constants';

/** interface for oAuth options */
export interface OauthOptions extends Options {
  userUri: string;
}

export type UnixTimestamp = number;
export interface OpensrpKeycloakTokenClaims {
  // expiration time (seconds since unix epoch)
  exp: UnixTimestamp;
  // issued at
  iat: UnixTimestamp;
  // time authentication occurred
  auth_time: UnixTimestamp;
  jti: string;
  iss: string;
  aud: string[];
  // subject
  sub: string;
  typ: string;
  azp: string;
  session_state: string;
  'allowed-origins': string[];
  realm_access: {
    roles: string[];
  };
  resource_access: {
    'realm-management': {
      roles: string[];
    };
    account: {
      roles: string[];
    };
  };
  scope: string;
  sid: string;
  email_verified: false;
  name: string;
  preferred_username: string;
  given_name: string;
  family_name: string;
}

export type RawOpensrpUserInfo = OpensrpKeycloakTokenClaims & {
  oAuth2Data: Record<string, any>;
};

/** interface for providers object */
export interface Providers {
  [key: string]: OauthOptions;
}

/** get oAuth provider from oAuthOptions
 * @param {OauthOptions} options - the options for the oAuth provider
 */
export function getProviderFromOptions(options: OauthOptions) {
  const { accessTokenUri, authorizationUri, clientId, redirectUri, scopes, state } = options;
  return new ClientOAuth2({
    accessTokenUri,
    authorizationUri,
    clientId,
    redirectUri,
    scopes,
    state
  });
}

/** Type definition for userInfo functions  */
export type UserInfoFnType = (obj: { [key: string]: any }) => SessionState | void;

/** Function to get onadata user info from api response object
 * @param {{[key: string]: any }} apiResponse - the API response object
 */
export function getOnadataUserInfo(apiResponse: { [key: string]: any }): SessionState | void {
  if (!apiResponse.username || !apiResponse.api_token) {
    throw new Error(OAUTH2_CALLBACK_ERROR);
  }
  return {
    authenticated: true,
    extraData: apiResponse,
    user: {
      email: apiResponse.email || '',
      gravatar: apiResponse.gravatar || '',
      name: apiResponse.name || '',
      username: apiResponse.username
    }
  };
}

/**
 * Takes current time and adds passes second
 * Incase passed seconds are invalid it returns null
 * @param {number} seconds - number of seconds
 */
const addSecToCurrentTime = (seconds: number, baseDate?: Date) => {
  const date = baseDate && !isNaN(baseDate.getTime()) ? baseDate : new Date(Date.now());
  return !isNaN(Number(seconds))
    ? new Date(date.setSeconds(date.getSeconds() + Number(seconds))).toISOString()
    : null;
};

/** Function to get OpenSRP user info from api response object
 * @param {RawOpensrpUserInfo} apiResponse - the API response object
 */
export function getOpenSRPUserInfo(apiRes: RawOpensrpUserInfo): SessionState {
  const {
    email_verified,
    oAuth2Data,
    given_name,
    family_name,
    preferred_username,
    realm_access,
    sub,
    name
  } = apiRes;
  const apiResponse = {
    roles: (realm_access?.roles ?? []).map((role: string) => `ROLE_${role}`),
    email: null,
    username: preferred_username,
    user_id: sub,
    preferred_name: name,
    family_name,
    given_name,
    email_verified,
    oAuth2Data
  };
  if (!apiResponse.username) {
    throw new Error(OAUTH2_CALLBACK_ERROR);
  }
  let responseCopy = { ...apiResponse };
  if (apiResponse.oAuth2Data) {
    const { expires_in, refresh_expires_in } = apiResponse.oAuth2Data;
    const authTime = new Date(apiRes.auth_time * 1000);
    const tokenExpiryTime = addSecToCurrentTime(expires_in, authTime);
    const refreshExpiryTime = addSecToCurrentTime(refresh_expires_in, authTime);
    responseCopy = {
      ...responseCopy,
      oAuth2Data: {
        ...apiResponse.oAuth2Data,
        ...(tokenExpiryTime && { token_expires_at: tokenExpiryTime }),
        ...(refreshExpiryTime && { refresh_expires_at: refreshExpiryTime })
      }
    };
  }
  return {
    authenticated: true,
    extraData: responseCopy,
    user: {
      email: '',
      gravatar: '',
      name: '',
      username: apiResponse.username
    }
  };
}
