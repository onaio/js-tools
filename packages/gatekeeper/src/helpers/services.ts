import {
  AuthenticateAction,
  authenticateUser,
  LogOutAction,
  logOutUser
} from '@onaio/session-reducer';
import ClientOAuth2 from 'client-oauth2';
import { ActionCreator, AnyAction, Dispatch } from 'redux';
import {
  authenticationProgress,
  AuthenticationProgressAction,
  RecordAction,
  recordResult
} from '../ducks/gatekeeper';
import { GENERIC_ERROR, OAUTH2_HTTP_ERROR, TOKEN_REFRESH_FAILED } from './constants';
import { getOnadataUserInfo, UserInfoFnType } from './oauth';
import { ErrorCallback, errorCallback } from './utils';

/** allowed http methods */
type HTTPMethod = 'GET' | 'POST' | 'get' | 'post';

/** Calls the oAuth provider to get user details
 * @param {URL} urlObject - the URL object defined by the location hash and href
 * @param {string} url - the URL that returns the user information for the provider
 * @param {ClientOAuth2} provider - the Oauth client object for the provider
 * @param {UserInfoFnType} userInfoCallback - function the gets user info from API response
 * @param {string} method - the HTTP method to use
 */
export async function oauth2Callback(
  urlObject: URL,
  url: string,
  provider: ClientOAuth2,
  userInfoCallback: UserInfoFnType,
  method: HTTPMethod = 'GET'
) {
  return provider.token.getToken(urlObject).then(async (oAuthObject: any) => {
    const response = await fetch(
      url,
      oAuthObject.sign({
        method,
        url
      })
    );

    if (!response.ok) {
      throw new Error(`${OAUTH2_HTTP_ERROR} ${response.status}`);
    }

    const data = await response.json();

    /** Add Oauth2 stuff */
    data.oAuth2Data = oAuthObject.data;

    return userInfoCallback(data);
  });
}

/** This function is used to fetch the user logging in by calling oauth2Callback
 * and then calling authenticateUser to store the user in the Redux store.
 * @param {URL} urlObject - the URL object defined by the location hash and href
 * @param {string} url - the URL that returns the user information for the provider
 * @param {ClientOAuth2} provider - the Oauth client object for the provider
 * @param {ActionCreator<AuthenticateAction>} authenticateActionCreator - the authenticate action creator function
 * @param {ActionCreator<RecordAction>} recordResultActionCreator - the recordResult action creator function
 * @param {UserInfoFnType} userInfoCallback - function the gets user info from API response
 * @param {ErrorCallback} errorCallbackFn - a function that handles error messages
 * @param {string} method - the HTTP method to use
 */
export async function fetchUser(
  urlObject: URL,
  resourceUrl: string,
  provider: ClientOAuth2,
  authenticateActionCreator: ActionCreator<AuthenticateAction> = authenticateUser,
  recordResultActionCreator: ActionCreator<RecordAction> = recordResult,
  userInfoCallback: UserInfoFnType = getOnadataUserInfo,
  errorCallbackFn: ErrorCallback = errorCallback,
  method: HTTPMethod = 'GET'
) {
  try {
    const responseInfo = await oauth2Callback(
      urlObject,
      resourceUrl,
      provider,
      userInfoCallback,
      method
    );
    if (responseInfo) {
      const { authenticated, user, extraData } = responseInfo;
      authenticateActionCreator(authenticated, user, extraData);
      recordResultActionCreator(true, extraData);
    } else {
      recordResultActionCreator(false, { error: GENERIC_ERROR });
      errorCallbackFn(GENERIC_ERROR);
    }
  } catch (error) {
    recordResultActionCreator(false, { error });
    errorCallbackFn(error.message);
  }
}

/** describes options to be passed to fetchState as second argument */
interface FetchStateActionCreators {
  authenticateActionCreator?: ActionCreator<AuthenticateAction>;
  recordResultActionCreator?: ActionCreator<RecordAction>;
  authenticationProgressCreator?: ActionCreator<AuthenticationProgressAction>;
  errorCallbackFn?: ErrorCallback;
  logoutActionCreator?: ActionCreator<LogOutAction>;
}

/** fetches session info from provided url
 * @params {string} url - points to location of the sessions
 * @params {options} - actionCreators
 */
export const fetchState = async (
  url: string,
  {
    authenticateActionCreator = authenticateUser,
    recordResultActionCreator = recordResult,
    authenticationProgressCreator = authenticationProgress,
    errorCallbackFn = errorCallback,
    logoutActionCreator = logOutUser
  }: FetchStateActionCreators
) => {
  authenticationProgressCreator(true);
  fetch(url)
    .then(res => {
      if (res.ok) {
        return res.json();
      } else {
        authenticationProgressCreator(false);
        throw new Error('fetching state failed');
      }
    })
    .then(data => {
      const { session } = data;
      if (!session) {
        logoutActionCreator();
        throw new Error('User is logged out');
      }
      const { authenticated, user, extraData } = session;
      authenticateActionCreator(authenticated, user, extraData);
      recordResultActionCreator(true, extraData);
      authenticationProgressCreator(false);
    })
    .catch(err => {
      recordResultActionCreator(false, { err });
      authenticationProgressCreator(false);
      errorCallbackFn(err);
    });
};

/** describes options to be passed to refreshToken as third argument */
interface RefreshTokenActionCreators {
  authenticateActionCreator?: ActionCreator<AuthenticateAction>;
  recordResultActionCreator?: ActionCreator<RecordAction>;
  errorCallbackFn?: ErrorCallback;
}

/**
 * call express API to Refresh token and return new token
 * @param {string} url - token refresh endpoint
 * @param {Dispatch<AnyAction>} dispatch - dispatch action
 * @param {RefreshTokenActionCreators} options - optional params
 */
export const refreshToken = async (
  url: string,
  dispatch: Dispatch<AnyAction>,
  {
    authenticateActionCreator = authenticateUser,
    errorCallbackFn = errorCallback,
    recordResultActionCreator = recordResult
  }: RefreshTokenActionCreators
) => {
  return fetch(url)
    .then(res => {
      if (res.ok) {
        return res.json();
      } else {
        throw new Error(TOKEN_REFRESH_FAILED);
      }
    })
    .then(data => {
      const { session } = data;
      if (!session) {
        throw new Error(TOKEN_REFRESH_FAILED);
      }
      const { authenticated, user, extraData } = session;
      const access_token = extraData?.oAuth2Data?.access_token;
      dispatch(authenticateActionCreator(authenticated, user, extraData));
      dispatch(recordResultActionCreator(true, extraData));
      return access_token;
    })
    .catch(err => {
      errorCallbackFn(err);
    });
};
