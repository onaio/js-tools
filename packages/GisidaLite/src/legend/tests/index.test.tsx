import { mount, shallow } from 'enzyme';
import React from 'react';
import { Legend } from '../../legend';
import { legendBlockBuilder } from '../../legend/utils/helpers';

describe('Legend Component', () => {
  it('renders without crashing', () => {
    shallow(<Legend />);
  });
  it('renders props correctly', () => {
    const wrapper = mount(<Legend />);
    expect(wrapper).toMatchSnapshot('legend props');
  });

  it('returns legend markup', () => {
    const wrapper = mount(<Legend />);
    expect(wrapper.html()).toMatchSnapshot('legend markup');
  });
  it('returns legend text correctly', () => {
    const wrapper = mount(<Legend />);
    expect(wrapper.text()).toMatchInlineSnapshot(`"01-23-66-1112-16Age break down"`);
  });
  it('builds legendblocks correctly', () => {
    expect(legendBlockBuilder('#000', '20', 'a-b')).toMatchInlineSnapshot(`
      <li
        style={
          Object {
            "background": "#000",
            "width": "20%",
          }
        }
      >
        a-b
      </li>
    `);
  });
});
