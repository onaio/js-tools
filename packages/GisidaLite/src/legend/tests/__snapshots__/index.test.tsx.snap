// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Legend Component renders props correctly: legend props 1`] = `ReactWrapper {}`;

exports[`Legend Component returns legend markup: legend markup 1`] = `"<div class=\\"legend\\"><div class=\\"legend-fill legend-label\\"><ul id=\\"legend-background\\"><li style=\\"background: rgb(189, 215, 231);\\">0</li><li style=\\"background: rgb(66, 146, 198);\\">1-2</li><li style=\\"background: rgb(33, 113, 181);\\">3-6</li><li style=\\"background: rgb(8, 81, 156);\\">6-11</li><li style=\\"background: rgb(8, 48, 107);\\">12-16</li></ul></div><span>Age break down</span></div>"`;
